internal/syncutil/once.go

```diff
@@ -68,3 +72,31 @@ func (o *Once) Do(ctx context.Context, f func() (interface{}, error)) (bool, int
 		}
 	}
 }
+
+// OnceOrRetry is an object that will perform exactly one success action.
+type OnceOrRetry struct {
+	done atomic.Bool
+	lock sync.Mutex
+}
+
+// OnceOrRetry calls the function f if and only if Do is being called for the
+// first time for this instance of Once or all previous calls to Do are failed.
+func (o *OnceOrRetry) Do(f func() error) error {
+	// fast path
+	if o.done.Load() {
+		return nil
+	}
+
+	// slow path
+	o.lock.Lock()
+	defer o.lock.Unlock()
+
+	if o.done.Load() {
+		return nil
+	}
+	if err := f(); err != nil {
+		return err
+	}
```

Comment:
This is cool!
nit: The name `OnceOrRetry` might be confusing as it does not convey the idea that the function will only run once successful. How about something like `OnceSuccess`?🤔

Reply:
SGTM

Reply:
There is a problem that `OnceSuccess` is not correct in grammar. On the other hand, `OnceSuccess` has ambiguity on the return value of the subsequent calls after the success. It could mean "succeed once and fail later".

Reply:
OK. It's an internal struct so never mind😂